File: examples/graph500/src/main.rs
use raphtory::{
    algorithms::{
        centrality::pagerank::unweighted_page_rank, components::weakly_connected_components,
    },
    arrow::{
        graph_impl::{ArrowGraph, ParquetLayerCols},
        query::{ast::Query, executors::rayon2, state::NoState, ForwardState, NodeSource},
    },
    core::entities::VID,
    prelude::GraphViewOps,
};
use raphtory_arrow::algorithms::connected_components;
use raphtory_storage::lanl::*;
use std::{any::Any, env, io::Write, num::NonZeroUsize, vec};

fn main() {
    // Retrieve named parameters from environment variables
    let resources_dir =
        env::var("resources_dir").expect("Environment variable 'resources_dir' not set");
    let target_dir = env::var("target_dir").expect("Environment variable 'target_dir' not set");

    // Set default values and then try to override them with environment variables if they exist
    let chunk_size: usize = env::var("chunk_size")
        .unwrap_or_else(|_| "268435456".to_string())
        .parse()
        .expect("Invalid value for 'chunk_size'");
    let t_props_chunk_size: usize = env::var("t_props_chunk_size")
        .unwrap_or_else(|_| "20000000".to_string())
        .parse()
        .expect("Invalid value for 't_props_chunk_size'");
    let read_chunk_size: usize = env::var("read_chunk_size")
        .unwrap_or_else(|_| "4000000".to_string())
        .parse()
        .expect("Invalid value for 'read_chunk_size'");
    let concurrent_files: usize = env::var("concurrent_files")
        .unwrap_or_else(|_| "1".to_string())
        .parse()
        .expect("Invalid value for 'concurrent_files'");
    let num_threads: usize = env::var("num_threads")
        .unwrap_or_else(|_| NonZeroUsize::new(1).unwrap().to_string())
        .parse()
        .expect("Invalid value for 'num_threads'");

    println!("Resources directory: {}", resources_dir);
    println!("Target directory: {}", target_dir);
    println!("Chunk Size: {}", chunk_size);
    println!("T Props Chunk Size: {}", t_props_chunk_size);
    println!("Read Chunk Size: {}", read_chunk_size);
    println!("Concurrent Files: {}", concurrent_files);
    println!("Num Threads: {}", num_threads);

    let graph_dir: &str = &target_dir.to_string();

    let parquet_dirs = vec![format!("{}", resources_dir)];

    let layer_parquet_cols: Vec<ParquetLayerCols> = vec![ParquetLayerCols {
        parquet_dir: &parquet_dirs[0],
        layer: "default",
        src_col: "src",
        dst_col: "dst",
        time_col: "time",
    }];

    let graph = match measure_without_print_results("Graph load from dir", || {
        ArrowGraph::load_from_dir(graph_dir)
    }) {
        Ok(g) => g,
        Err(e) => {
            println!("Failed to load saved graph. Attempting to load from parquet files...");
            measure_without_print_results("Graph load from parquets", || {
                ArrowGraph::load_from_parquets(
                    graph_dir,
                    layer_parquet_cols,
                    None,
                    chunk_size,
                    t_props_chunk_size,
                    Some(read_chunk_size),
                    Some(concurrent_files),
                    num_threads,
                )
            })
            .expect("Failed to load the graph from parquet files")
        }
    };

    println!("Node count = {}", graph.count_nodes());
    println!("Edge count = {}", graph.count_edges());
    println!("Earliest time = {}", graph.earliest_time().unwrap());
    println!("Latest time = {}", graph.latest_time().unwrap());

    fn iterate_all_edges(g: &ArrowGraph) {
        for e in g.edges() {
            e.edge.src().0;
        }
    }

    measure_without_print_results("Iterate All Edges", || iterate_all_edges(&graph));

    measure_without_print_results("CC", || {
        connected_components::connected_components(graph.as_ref())
    });
    measure_without_print_results("Weakly CC", || {
        weakly_connected_components(&graph, 20, None)
    });
    measure_without_print_results("Page Rank", || {
        unweighted_page_rank(&graph, Some(100), Some(1), None, true, None)
    });

    measure_without_print_results("Static Multi Hop Query (1 node)", || {
        static_multi_hop_single_node_query(&graph)
    });
    measure_without_print_results("Static Multi Hop Query (100 node)", || {
        static_multi_hop_multi_node_query(&graph)
    });
    measure_without_print_results("Temporal Multi Hop Query (1 node)", || {
        multi_hop_single_node_query(&graph)
    });
    measure_without_print_results("Temporal Multi Hop Query (100 node)", || {
        multi_hop_multi_node_query(&graph)
    });
}

fn static_multi_hop_single_node_query(graph: &ArrowGraph) {
    let nodes = vec![VID(3000)];

    let (sender, receiver) = std::sync::mpsc::channel();
    let query = Query::new().out("default").out("default").channel([sender]);

    let result = rayon2::execute::<NoState>(query, NodeSource::NodeIds(nodes), &graph, |_| NoState);
    assert!(result.is_ok());

    let _ = receiver.into_iter().collect::<Vec<_>>();
}

fn static_multi_hop_multi_node_query(graph: &ArrowGraph) {
    let nodes: Vec<VID> = (0..100).map(VID).collect();

    let (sender, receiver) = std::sync::mpsc::channel();
    let query = Query::new()
        .out_limit("default", 100)
        .out_limit("default", 100)
        .out_limit("default", 100)
        .out_limit("default", 100)
        .channel([sender]);

    let result = rayon2::execute::<NoState>(query, NodeSource::NodeIds(nodes), &graph, |_| {
        NoState::new()
    });
    assert!(result.is_ok());

    match receiver.recv() {
        Ok((_, vid)) => {
            println!("VID {}", vid.0);
        }
        Err(e) => {
            println!("Error receiving result: {:?}", e);
        }
    }
}

fn multi_hop_single_node_query(graph: &ArrowGraph) {
    let nodes = vec![VID(0)];

    let query: Query<ForwardState> = Query::new()
        .out_limit("default", 100)
        .out_limit("default", 100)
        .out_limit("default", 100)
        .out_limit("default", 100)
        .path("hop", |mut writer, state: ForwardState| {
            serde_json::to_writer(&mut writer, &state.path).unwrap();
            write!(writer, "\n").unwrap();
        });

    let result =
        rayon2::execute::<ForwardState>(query, NodeSource::NodeIds(nodes), graph, |node| {
            let earliest = node.earliest();
            ForwardState::at_time(node, earliest, 100)
        });

    assert!(result.is_ok());
}

fn multi_hop_multi_node_query(graph: &ArrowGraph) {
    let nodes: Vec<VID> = (0..100).map(VID).collect();

    let query: Query<ForwardState> = Query::new()
        .out_limit("default", 100)
        .out_limit("default", 100)
        .out_limit("default", 100)
        .out_limit("default", 100)
        .path("hop", |mut writer, state: ForwardState| {
            serde_json::to_writer(&mut writer, &state.path).unwrap();
            write!(writer, "\n").unwrap();
        });

    let result =
        rayon2::execute::<ForwardState>(query, NodeSource::NodeIds(nodes), graph, |node| {
            let earliest = node.earliest();
            ForwardState::at_time(node, earliest, 100)
        });

    assert!(result.is_ok());
}

File: examples/graph500/python/main.py
import argparse
from raphtory import ArrowGraph
from raphtory import algorithms
import time
from typing import TypeVar, Callable
from tqdm import tqdm
import os
import sys
from raphtory import ArrowGraph, Query, State


B = TypeVar('B')

def measure(name: str, f: Callable[..., B], *args, print_result: bool = True) -> B:
    start_time = time.time()
    result = f(*args)
    elapsed_time = time.time() - start_time
    
    time_unit = "s"
    elapsed_time_display = elapsed_time
    if elapsed_time < 1:
        time_unit = "ms"
        elapsed_time_display *= 1000

    if print_result:
        print(f"Running {name}: time: {elapsed_time_display:.3f}{time_unit}, result: {result}")
    else:
        print(f"Running {name}: time: {elapsed_time_display:.3f}{time_unit}")

    return result

def hop_queries(graph, nodes, layer, hops, temporal, start=None, keep_path=False):
    q = Query.from_node_ids(nodes)
    
    for _ in range(hops):
        q = q.out(layer)
    
    if temporal:
        max_duration = 2**31 - 1
        state = State.path_window(keep_path=keep_path, start_t=start, duration=max_duration)
    else:
        state = State.no_state()
    
    q.run(graph, state)

def static_multi_hop_single_node_query(graph: ArrowGraph):
    hop_queries(graph, [0], "default", 3, False)

def static_multi_hop_multi_node_query(graph: ArrowGraph):
    nodes = list(range(100))
    hop_queries(graph, nodes, "default", 3, False)
    
def multi_hop_single_node_query(graph: ArrowGraph):
    hop_queries(graph, [0], "default", 3, True, 10, True)

def multi_hop_multi_node_query(graph: ArrowGraph):
    nodes = list(range(100))
    hop_queries(graph, nodes, "default", 3, True, 10, True)

def main(graph_dir, resources_dir, chunk_size, t_props_chunk_size, read_chunk_size, concurrent_files, num_threads):

    if resources_dir is None or graph_dir is None:
        raise ValueError("Both 'resources_dir' and 'target_dir' environment variables must be set.")

    print(f"Resources directory: {resources_dir}")
    print(f"Target directory: {graph_dir}")
    print(f"Chunk size: {chunk_size}")
    print(f"t_props chunk size: {t_props_chunk_size}")
    print(f"Read chunk size: {read_chunk_size}")
    print(f"Concurrent files: {concurrent_files}")
    print(f"Number of threads: {num_threads}")

    layer_parquet_cols = [
        {
            "parquet_dir": resources_dir,
            "layer": "default",
            "src_col": "src",
            "dst_col": "dst",
            "time_col": "time",
        }
    ]

    # # Read the Parquet file
    # table = pq.read_table(parquet_dir + '/part-00000-8b31eaa4-2bd9-4f07-b61c-a353aed2af22-c000.snappy.parquet')
    # print(table.schema)

    print()
    try:
        g = measure("Graph load from dir", ArrowGraph.load_from_dir, graph_dir, print_result=False)
    except Exception as e:
        g = measure(
            "Graph load from parquets", 
            ArrowGraph.load_from_parquets,
            graph_dir,
            layer_parquet_cols,
            None,
            chunk_size, 
            t_props_chunk_size,
            read_chunk_size,
            concurrent_files,
            num_threads,
            print_result=False
        )

    print("Nodes count =", g.count_nodes())
    print("Edges count =", g.count_edges())
    print("Earliest time =", g.earliest_time)
    print("Latest time =", g.latest_time)
    
    measure("Static Multi Hop Single Node Query", static_multi_hop_single_node_query, g, print_result=False)
    measure("Static Multi Hop Multi Node Query", static_multi_hop_multi_node_query, g, print_result=False)
    measure("Temporal Multi Hop Singl Node Query", multi_hop_single_node_query, g, print_result=False)
    measure("Temporal Multi Hop Multi Node Query", multi_hop_multi_node_query, g, print_result=False)
    measure("CC", algorithms.connected_components, g, print_result=False)
    measure("Page Rank", algorithms.pagerank, g, 100, print_result=False)

    
if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Run LANL example queries')
    parser.add_argument('--graph-dir', type=str, help='The directory of the graph')
    parser.add_argument('--resources-dir', type=str, help='paths to the parquet directory')
    parser.add_argument('--chunk-size', type=int, default=268435456, help='Chunk size')
    parser.add_argument('--t-props-chunk-size', type=int, default=20000000, help='t_props chunk size')
    parser.add_argument('--read-chunk-size', type=int, default=4000000, help='Read chunk size')
    parser.add_argument('--concurrent-files', type=int, default=1, help='Concurrent files')
    parser.add_argument('--num-threads', type=int, default=4, help='Number of threads')
    args = parser.parse_args()
    main(args.graph_dir, args.resources_dir, args.chunk_size, args.t_props_chunk_size, args.read_chunk_size, args.concurrent_files, args.num_threads)
